core/src/main/java/com/sequenceiq/cloudbreak/service/environment/EnvironmentService.java

```diff
@@ -400,11 +401,19 @@ public DetailedEnvironmentV4Response registerExternalDatalake(String environment
 
     @Override
     protected void prepareDeletion(Environment environment) {
-        Long aliveStacks = stackService.countAliveByEnvironment(environment);
-        Long aliveClusters = clusterService.countAliveByEnvironment(environment);
-        if (aliveStacks > 0 || aliveClusters > 0) {
+        List<String> aliveStacks = stackService.getNameOfAliveByEnvironment(environment);
+        if (!aliveStacks.isEmpty()) {
             throw new BadRequestException("Cannot delete environment. "
-                    + "All clusters must be terminated before environment deletion. Alive clusters: " + aliveStacks);
+                    + "All clusters must be terminated before environment deletion. Alive cluster(s): "
+                    + String.join(", ", aliveStacks) + "."
+            );
+        }
+        List<String> aliveClusters = clusterService.getNameOfAliveByEnvironment(environment);
+        if (!aliveClusters.isEmpty()) {
+            throw new BadRequestException("Cannot delete environment. "
+                    + "All clusters must be terminated before environment deletion. One or more clusters are in bad state after termination: "
```

Comment:
Are you sure at this point of the execution that all the clusters are in `bad state after termination` and non of them simply alive?

Reply:
Well, stack is in DELETE_COMPLETED: that is, the cloud instances are already gone. 
There have been such errors already, though not very common. A scenario for that was: delete a cluster that is in UPDATE_IN_PROGRESS state. Sometimes the DELETE_COMPLETED status was overwritten with UPDATE_IN_PROGRESS.

Reply:
Thanks for the explanation. Would you mind calling it `invalid state` instead of `bad state`? That is my only request.

Reply:
done, thanks!